File: android/app/src/main/java/vn/team/freechat/adapter/MessageListAdapter.java
package vn.team.freechat.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.tvd12.ezyfoxserver.client.constant.EzyConstant;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import tvd12.com.ezyfoxserver.client.R;
import vn.team.freechat.data.Message;
import vn.team.freechat.data.MessageReceived;

/**
 * Created by tavandung12 on 10/5/18.
 */

public class MessageListAdapter extends RecyclerView.Adapter {

    private final Context context;
    private final List<Message> items;

    private static final int VIEW_TYPE_MESSAGE_SEDING = 1;
    private static final int VIEW_TYPE_MESSAGE_SENT = 2;
    private static final int VIEW_TYPE_MESSAGE_RECEIVED = 3;

    public MessageListAdapter(Context context) {
        this(context, new ArrayList<Message>());
    }

    public MessageListAdapter(Context context, List<Message> messages) {
        this.context = context;
        this.items = messages;
    }

    public void addItem(Message item) {
        this.items.add(item);
    }

    public void addItems(Collection<Message> items) {
        this.items.addAll(items);
    }

    @Override
    public MessageHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = null;
        switch (viewType) {
            case VIEW_TYPE_MESSAGE_SENT:
                view = LayoutInflater.from(context)
                        .inflate(R.layout.component_message_item_sent, parent, false);
                return new MessageSentHolder(view);
            case VIEW_TYPE_MESSAGE_RECEIVED:
                view = LayoutInflater.from(context)
                        .inflate(R.layout.component_message_item_received, parent, false);
                return new MessageReceivedHolder(view);
            default:
                throw new IllegalArgumentException("has no view with type: " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Message message = getItem(position);
        ((MessageHolder)holder).bind(message);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public Message getItem(int position) {
        Message item = items.get(position);
        return item;
    }

    @Override
    public int getItemViewType(int position) {
        Message item = getItem(position);
        EzyConstant type = item.getType();
        return type.getId();
    }
}

class MessageHolder<T extends Message> extends RecyclerView.ViewHolder {

    private TextView messageView;
    private TextView timeView;


    public MessageHolder(View itemView) {
        super(itemView);
        this.messageView = itemView.findViewById(R.id.message);
        this.timeView = itemView.findViewById(R.id.time);
    }

    public void bind(T message) {
        this.messageView.setText(message.getMessage());
        this.timeView.setText(message.getSentTimeString());
    }
}

class MessageSentHolder extends MessageHolder {

    public MessageSentHolder(View itemView) {
        super(itemView);
    }
}

class MessageReceivedHolder extends MessageHolder<MessageReceived> {

    private ImageView avatarView;
    private TextView fromView;


    public MessageReceivedHolder(View itemView) {
        super(itemView);
        this.avatarView = itemView.findViewById(R.id.avatar);
        this.fromView = itemView.findViewById(R.id.from);
    }

    @Override
    public void bind(MessageReceived message) {
        super.bind(message);
        fromView.setText(message.getFrom());
    }
}

File: kotlin/app/src/main/java/vn/team/freechat_kotlin/adapter/MessageListAdapter.kt
package vn.team.freechat_kotlin.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import vn.team.freechat_kotlin.R
import vn.team.freechat_kotlin.data.Message
import vn.team.freechat_kotlin.data.MessageReceived
import vn.team.freechat_kotlin.data.MessageSent
import java.util.*
import java.util.Collection


/**
 * Created by tavandung12 on 10/5/18.
 */

class MessageListAdapter : RecyclerView.Adapter<MessageHolder<Message>> {

    private val context : Context
    private val inflater : LayoutInflater
    private val items : MutableList<Message>

    companion object {
        private val VIEW_TYPE_MESSAGE_SEDING = 1
        private val VIEW_TYPE_MESSAGE_SENT = 2
        private val VIEW_TYPE_MESSAGE_RECEIVED = 3
    }

    constructor(context: Context) : super() {
        this.context = context
        this.inflater = LayoutInflater.from(context)
        this.items = ArrayList<Message>()
    }

    fun addItem(item: Message) {
        this.items.add(item)
    }

    fun addItems(items: Collection<Message>) {
        this.items.addAll(items)
    }

    override fun onCreateViewHolder(parent : ViewGroup?, viewType: Int) : MessageHolder<Message> {
        var view : View?
        return when (viewType) {
            VIEW_TYPE_MESSAGE_SENT -> {
                view = inflater.inflate(R.layout.component_message_item_sent, parent, false)
                MessageSentHolder(view) as MessageHolder<Message>
            }
            VIEW_TYPE_MESSAGE_RECEIVED -> {
                view = inflater.inflate(R.layout.component_message_item_received, parent, false)
                MessageReceivedHolder(view) as MessageHolder<Message>
            }
            else -> {
                throw IllegalArgumentException ("has no view with type: " + viewType)
            }
        }
    }

    override fun onBindViewHolder(holder: MessageHolder<Message>?, position: Int) {
        val message = getItem(position)
        (holder as MessageHolder<Message>).bind(message)
    }

    override fun getItemCount() : Int {
        return items.size
    }

    private fun getItem(position: Int) : Message {
        val item = items[position]
        return item
    }

    override fun getItemViewType(position: Int): Int {
        val item = getItem(position)
        val type = item.getType()
        return type.id
    }
}

open class MessageHolder<E> : RecyclerView.ViewHolder {

    private val messageView : TextView
    private val timeView : TextView

    constructor(itemView: View) : super(itemView) {
        this.messageView = itemView.findViewById(R.id.message)
        this.timeView = itemView.findViewById(R.id.time)
    }

    open fun bind(message: E) {
        val m = message as Message
        this.messageView.text = m.message;
        this.timeView.text = m.getSendTimeString()
    }
}

class MessageSentHolder : MessageHolder<MessageSent> {

    constructor(itemView: View) : super(itemView)
}

class MessageReceivedHolder : MessageHolder<MessageReceived> {

    private var avatarView: ImageView? = null
    private var fromView: TextView? = null


    constructor(itemView: View) : super(itemView) {
        this.avatarView = itemView.findViewById(R.id.avatar)
        this.fromView = itemView.findViewById(R.id.from)
    }

    override fun bind(message: MessageReceived) {
        super.bind(message)
        fromView?.text = message.from
    }
}